File: solutions/_783_minimum_distance_between_bst_nodes/Solution.java
package _783_minimum_distance_between_bst_nodes;

class Solution {

  private int minDiff = Integer.MAX_VALUE;
  private int prev = -1;

  public int minDiffInBST(TreeNode root) {
    inOrder(root);
    return minDiff;
  }

  private void inOrder(TreeNode node) {
    if (node == null) return;

    inOrder(node.left);

    if (prev != -1) {
      minDiff = Math.min(minDiff, node.val - prev);
    }

    prev = node.val;
    inOrder(node.right);
  }
}
/*
https://leetcode.com/problems/minimum-distance-between-bst-nodes/
*/

File: solutions/_783_minimum_distance_between_bst_nodes/solution.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.minDiff = float("inf")
        self.prev = None

    def inOrder(self, node: Optional[TreeNode]):
        if not node:
            return

        self.inOrder(node.left)
        if self.prev is not None:
            self.minDiff = min(self.minDiff, node.val - self.prev)

        self.prev = node.val
        self.inOrder(node.right)

    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        self.inOrder(root)
        return self.minDiff


# https://leetcode.com/problems/minimum-distance-between-bst-nodes/
